### PR TITLE
Support for WASI threads

### PR DESCRIPTION
**Pull request by Milek7**
wasi-libc provides pthread implementation, so use that. This way it behaves almost like POSIX-like platform (it does not support signals). Unfortunately macrology does get pretty complicated, because it is WASM, but does not need to use most existing WASM support code but generic POSIX instead.

wasi-sdk is updated to version 21 for https://github.com/WebAssembly/wasi-sdk/commit/c891cd2036d5630e67826f70aec9c0ba117c67e9 and https://github.com/WebAssembly/wasi-libc/commit/ec4566beae84e54952637f0bf61bee4b4cacc087

`pthread_exit` does not exist, but returning from `start_wrapper` seems to work fine, not sure when other ways of exiting a thread are invoked.

Conditionals in csproj are patched to work correctly with WASI threads, but I'm not really sure whether `FeatureWasmManagedThreads` should be defined at all for this configuration. It's not a browser, has POSIX threads, so this whole complication with ifdefing unsupported browser attributes and adding special assemblies for `WebAssembly.Threading` seems completely unnecessary. I don't know how to fix this cleanly, though.

When testing on wasmtime it needs https://github.com/bytecodealliance/wasmtime/pull/7884

### CODE REVIEW

**Comment by Milek7** [comment]
@dotnet-policy-service agree company="SimRail S.A."

File: src/mono/CMakeLists.txt [hunk 1]
```diff
@@ -183,11 +187,17 @@ endif()
...
+if("${triple}" STREQUAL "wasm32-wasi-threads")
+  set(CMAKE_SYSTEM_VARIANT "threads")
+endif()
+
 message ("CMAKE_SYSTEM_NAME=${CMAKE_SYSTEM_NAME}")
 message ("CMAKE_SYSTEM_VARIANT=${CMAKE_SYSTEM_VARIANT}")
 
 set(CLR_CMAKE_HOST_OS ${CMAKE_SYSTEM_NAME})
-string(TOLOWER ${CLR_CMAKE_HOST_OS} CLR_CMAKE_HOST_OS)
+set(CLR_CMAKE_HOST_VARIANT ${CMAKE_SYSTEM_VARIANT})
```
Comment (kg) [inline src/mono/CMakeLists.txt:198]: i'm not super familiar with cmake, but it seems weird that the *host* variant is either threads or not threads depending on the target you're building for. don't all cmake build hosts have threads?

Comment (Milek7) [reply src/mono/CMakeLists.txt:198]: I think this is referring not to build host but mono runtime host. (as opposed to target system for which runtime will do codegen)
Still, I'm not sure how that interacting with `DISABLE_THREADS` (shouldn't there be TARGET/HOST_DISABLE_THREADS?), but in any case it was used for wasm this way before.

File: src/mono/mono/utils/mono-threads-wasm.c [hunk 2]
```diff
@@ -593,7 +593,113 @@ mono_threads_wasm_sync_run_in_target_thread_vii (pthread_t target_thread, void (
 
 #endif /* DISABLE_THREADS */
 
-#endif /* HOST_BROWSER */
+#endif /* defined(USE_WASM_BACKEND) */
+
+#elif defined(USE_PTHREAD_WASM_BACKEND)
+
+struct __pthread {
+	struct pthread *self;
```
[the remark this answers is not in the document]
Comment (Milek7) [reply src/mono/mono/utils/mono-threads-wasm.c:601]: I don't know how to get stack bounds otherwise, because wasi-libc doesn't have pthread_getattr_np.

**Comment by Milek7** [comment]
> This is WASIX extensions, right ? What would happen when we upgrade to WASI preview 2 ?

It uses pthreads from wasi-libc, which is currently implemented using wasi-threads proposal (https://github.com/WebAssembly/wasi-threads)
When wasi-libc switches underlying implementation it shouldn't require changes here, but as far I know component-model based preview2 isn't expected to support threads soon, so pthread-supporting toolchain will stay on preview1 for quite some time.

File: src/mono/mono/utils/mono-threads-wasm.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (Milek7) [reply src/mono/mono/utils/mono-threads-wasm.c:601]: https://github.com/WebAssembly/wasi-libc/pull/470

**Comment by pavelsavara** [comment]
@Milek7 we appreciate your efforts! 

I would like to share some context to set your and our expectations. As you know, the wasm threading is still developing and the final implementation may be very different from what you are trying to achieve in this PR.

Our view is that we want to have dotnet based on the standard component based WASI, rather than POSIX/WASIX extensions. On the other hand, as you pointed out, threads are not ready. 

We are willing to explore this, given that we all know this API would eventually go away. 
(The WASI flavor of dotnet is still experimental, probably until WASI itself reaches v1.0, we will not guarantee any compatibility or behavior)

As you probably noticed, our big focus at the moment is to [enable threads in the browser](https://github.com/dotnet/runtime/issues/85592) (with emscripten). Browser limitations as well as interop with JS requires that we work-around many concepts. Historically we only had Mono on emscripten and that combination was referred to as `wasm` in the codebase. About year ago we added `WASI` as another OS and recently made refactoring of the source code directories to reflect `browser` OS. But part of the `#ifdef` probably still reflects `wasm` meaning `wasm-browser`. Story of `FeatureWasmManagedThreads` in csproj is similar. It surely makes sense sort it out. This is all Mono, but now there is also [NativeAOT experimental branch](https://github.com/dotnet/runtimelab/tree/feature/NativeAOT-LLVM) of dotnet runtime on WASM.

For context, in the browser major features which influence threads are:
- you need to keep yielding to browser event loop 
    - otherwise lot of things don't work: networking, rendering, debugging, new thread creation ...
- you can't synchronously block on UI thread (roughly for same reasons)
- you can schedule yourself callback via `setTimeout`. 
    - This is the way how we overcome limitations above. 
    - It allows us to implement single-threaded "thread pool" and timer and await JS promises.
- Some of those limitations are lifted for multi-threaded (MT) browser, but there is JS object thread affinity.

We have following combinations ahead
- ST browser - stable
- MT browser - under heavy development
- ST WASI - possibly after we finish MT browser
- MT WASIX - we did not prioritize it in any way yet
- MT WASI - we are waiting for WASI component model

What need to be done for WASI
 - We need to conceptually figure out which of dotnet features we solved for (single-threaded) browser, could be solved for single-threaded WASI.
     - ST "thread pool" to schedule async `Task` to single thread.
     - Timers - there is no `setTimeout` in WASI
     - PLINQ
     - GC, finalizer
     - knowing this would inform us on the `#ifdef`
     - enable unit test coverage for all BCL. Right now we only have "hello world" smoke test running on CI.
 - make WASI MT with WASIX shared memory work
     - this may be side-step for us. 
     - implement thread-pool, timers, PLINQ, GC - this may be very close to normal unix. We don't need to yield to (browser) event loop etc.
     - enable unit tests for all BCL on MT, fix problems
 - adopt WASIp2 component model in dotnet PAL. 
     - Possibly call that API directly rather than via libc. Would wasm-threads libc dependency hold us back ?
     - This may bring networking. Browser HTTP and WS adapters are non-trivial. WASIp3 with futures and streams will change it again. Shall we wait for it to arrive ?
 - make WASI MT with component model thread
     - Is component based MT going to bring WASM linear memory shared across threads ?
     - Would there be [host "event loop"](https://github.com/WebAssembly/WASI/issues/582#issuecomment-1936125720) ? Because there will be WASIp3 futures.
     - it may mean that we will need to partially align it with browser MT and that WASIX MT will be complexity to get rid of, at that point.

Right now, we are not actively working on any of the above.

Are you still interested ? 
Which parts you would like to work on ? 
What's your motivation or use-case for working on this ?

**Comment by Milek7** [comment]
> What's your motivation or use-case for working on this ?

My motivation is running existing C# codebase in wasm runtime. As you noted, single-threaded requires yielding to caller, which requires changes to both library code and applications (which may assume that threads exist, use locks, and blocking functions). Multi-threading though makes it fit into posix-like model, which enables library and application code to work without changes.

> * make WASI MT with WASIX shared memory work

This PR uses wasi-threads proposal (as implemented in eg. wasmtime) and does not use wasmer WASIX extensions.

> * implement thread-pool, timers, PLINQ, GC - this may be very close to normal unix. We don't need to yield to (browser) event loop etc.

Indeed that's why I'm pursuing this approach. Admittedly I didn't do extensive testing yet, but this (Thread, ThreadPool, Task, etc.) should already work with this PR.

Given that it seems component model multithreading won't be available for quite some time (I have seen statements that "The Component Model does not have support for threads at this time and that's not expected to be supported until the time scale of years in the future"), I think it is reasonable to merge this approach for time being. When it comes I think it is expected that wasi-libc will still provide pthread implementation requiring no changes here. Targeting component model directly bypassing libc would likely require some changes, though I'm not exactly sure why you would do this, considering that in practice it would probably mean copy-pasting musl code. (unless it would also bring Wasm GC, but that would be complete architecture change anyway).

I can look into threading bugs if some will come up with more testing, and potentially untangle that `FeatureWasmManagedThreads` flag if there would be guidelines what to do with it. (like I said before, it works but feels superfluous to ifdef UnsupportedOS for browser when this isn't a browser anyway)

**Comment by pavelsavara** [comment]
> > What's your motivation or use-case for working on this ?
> My motivation is running existing C# codebase in wasm runtime

I'm interested in more details, if you are willing to share them. 
What makes wasm attractive compared with say "chiselled" docker ? (this is Q from wasm fan boy)

> I didn't do extensive testing yet, but this (Thread, ThreadPool, Task, etc.) should already work with this PR.

We can help enable WASI full library test suite for you on this PR. Is it ok for us to push into your branch ?

There are also some failing CI legs on this PR, because we will need to bump SDK in the Helix VM images too.

> potentially untangle that `FeatureWasmManagedThreads` flag 

We are discussing that with the team. TBH we don't know exactly what we want yet. Will keep you posted.

**Comment by vargaz** [comment]
Maybe the bump of the wasi sdk should be done in a separate PR ?

**Comment by Milek7** [comment]
> We can help enable WASI full library test suite for you on this PR. Is it ok for us to push into your branch ?

Thanks, feel free to do it.

> I'm interested in more details, if you are willing to share them.

It's probably rather unconventional use-case :) Actually we are embedding it in desktop application, for various reasons it is convenient to have it compiled as WASM module (no extra processes, easy state cleanup for restart).

**Comment by pavelsavara** [comment]
> PS: For testing with wasmtime it will need fresh build (for [bytecodealliance/wasmtime#7884](https://github.com/bytecodealliance/wasmtime/pull/7884))

Is there feed with nighty build ?

https://github.com/dotnet/runtime/blob/954ead3c0a37f8e8685dc11e19377caf975e9c8d/eng/testing/wasi-provisioning.targets#L21-L23

**Comment by Milek7** [comment]
https://github.com/bytecodealliance/wasmtime/releases/download/dev/wasmtime-dev-x86_64-linux.tar.xz

**Comment by vargaz** [comment]
The wasi sdk bump was merged.

**Comment by pavelsavara** [comment]
I didn't forget about it and it's still on my TODO list. I plan to cleanup the `ifdef` situation first, thanks for your patience.

**Comment by pavelsavara** [comment]
We don't have bandwidth to keep WASIp1 at the same time as we want to [switch to WASIp2](https://github.com/dotnet/runtime/pull/103752). It would require even more complex and conditional build scripts. I'm sorry.

**Pull request closed without merging**